brokers/vscode/broker.go

```diff
@@ -0,0 +1,296 @@
+			},
+		},
+	}
+	return &c
+}
+
+func (broker *VSCodeExtensionBroker) addPortToTooling(toolingConf *model.ToolingConf, pj *packageJSON) {
+	port := findPort()
+	sPort := strconv.Itoa(port)
+	endpointName := "port" + sPort
+	var re = regexp.MustCompile(`[^a-z_0-9]+`)
+	prettyID := re.ReplaceAllString(pj.Publisher+"_"+pj.Name, `_`)
+	theiaEnvVar1 := "THEIA_PLUGIN_REMOTE_ENDPOINT_" + prettyID
+	theiaEnvVarValue := "ws://" + endpointName + ":" + sPort
+
+	toolingConf.Containers[0].Ports = append(toolingConf.Containers[0].Ports, model.ExposedPort{ExposedPort: port})
+	toolingConf.Endpoints = append(toolingConf.Endpoints, model.Endpoint{
+		Name:       endpointName,
+		Public:     false,
+		TargetPort: port,
+	})
+	toolingConf.Containers[0].Env = append(toolingConf.Containers[0].Env, model.EnvVar{Name: "THEIA_PLUGIN_ENDPOINT_PORT", Value: sPort})
+	toolingConf.WorkspaceEnv = append(toolingConf.WorkspaceEnv, model.EnvVar{Name: theiaEnvVar1, Value: theiaEnvVarValue})
+}
+
+func randomNumberAsString() string {
+	port := findPort()
+	return strconv.Itoa(port)
+}
+
+func findPort() int {
+	return 4000 + rand.Intn(6000)
+}
+
+func (broker *VSCodeExtensionBroker) download(extension string, dest string, meta model.PluginMeta) error {
+	response, err := fetchExtensionInfo(extension, meta)
+	if err != nil {
+		return err
+	}
+
+	URL, err := findAssetURL(response, meta)
+	if err != nil {
+		return err
+	}
+
+	err = broker.ioUtil.Download(URL, dest)
+	return err
+}
+
+func fetchExtensionInfo(extension string, meta model.PluginMeta) ([]byte, error) {
+	re, err := regexp.Compile(`^vscode:extension/(.*)`)
+	if err != nil {
+		m := fmt.Sprintf("VS Code extension id '%s' parsing failed for plugin %s:%s", extension, meta.ID, meta.Version)
+		return nil, errors.New(m)
+	}
+	groups := re.FindStringSubmatch(extension)
+	if len(groups) != 2 {
+		m := fmt.Sprintf("VS Code extension id '%s' parsing failed for plugin %s:%s", extension, meta.ID, meta.Version)
+		return nil, errors.New(m)
+	}
+	extName := groups[1]
+	body := []byte(fmt.Sprintf(bodyFmt, extName))
+	req, err := http.NewRequest("POST", marketplace, bytes.NewBuffer(body))
+	req.Header.Set("Accept", "application/json;api-version=3.0-preview.1")
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		m := fmt.Sprintf("VS Code extension downloading failed %s:%s. Error: %s", meta.ID, meta.Version, err)
+		return nil, errors.New(m)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		m := fmt.Sprintf("VS Code extension downloading failed %s:%s. Status: %q", meta.ID, meta.Version, resp.StatusCode)
+		return nil, errors.New(m)
+	}
+
+	body, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		m := fmt.Sprintf("VS Code extension downloading failed %s:%s. Error: %s", meta.ID, meta.Version, err)
+		return nil, errors.New(m)
+	}
+
+	return body, nil
+}
+
+func findAssetURL(response []byte, meta model.PluginMeta) (string, error) {
```

Comment:
`meta` parameter is not used in this func, then why do we need it here?

Reply:
It is used now